src/codegen/operations.rs

```diff
@@ -2724,3 +2726,181 @@ fn codegen_mcopy<'c, 'r>(
+    let start_block = region.append_block(Block::new(&[]));
+    let context = &op_ctx.mlir_context;
+    let location = Location::unknown(context);
+    let uint256 = IntegerType::new(context, 256);
+    let uint8 = IntegerType::new(context, 8);
+    let uint1 = IntegerType::new(context, 1);
+    let ptr_type = pointer(context, 0);
+
+    // Check there's enough elements in stack
+    let flag = check_stack_has_at_least(context, &start_block, 1)?;
+    // Check there's enough gas
+    let gas_flag = consume_gas(context, &start_block, gas_cost::CALLDATALOAD)?;
+
+    let condition = start_block
+        .append_operation(arith::andi(gas_flag, flag, location))
+        .result(0)?
+        .into();
+
+    let ok_block = region.append_block(Block::new(&[]));
+
+    start_block.append_operation(cf::cond_br(
+        context,
+        condition,
+        &ok_block,
+        &op_ctx.revert_block,
+        &[],
+        &[],
+        location,
+    ));
+
+    let offset = stack_pop(context, &ok_block)?;
+
+    // TODO: add a calldata_ptr and size setup
+
+    let calldata_ptr = op_ctx.get_calldata_ptr_syscall(&ok_block, location)?;
+
+    // max_slice_width = 32
+    let max_slice_width = ok_block
+        .append_operation(arith::constant(
+            context,
+            integer_constant_from_i64(context, 32).into(),
+            location,
+        ))
+        .result(0)?
+        .into();
+
+    let calldata_size_u32 = op_ctx.get_calldata_size_syscall(&ok_block, location)?;
+    // convert calldata_size from u32 to u256
+    let calldata_size = ok_block
+        .append_operation(arith::extui(calldata_size_u32, uint256.into(), location))
+        .result(0)?
+        .into();
+
+    let zero = ok_block
+        .append_operation(arith::constant(
+            context,
+            IntegerAttribute::new(uint256.into(), 0).into(),
+            location,
+        ))
+        .result(0)?
+        .into();
+
+    let offset_ok_block = region.append_block(Block::new(&[]));
+    let offset_bad_block = region.append_block(Block::new(&[]));
+    let end_block = region.append_block(Block::new(&[]));
+
+    // offset < calldata_size =>  offset_ok
+    let offset_ok = ok_block
+        .append_operation(arith::cmpi(
+            context,
+            arith::CmpiPredicate::Ult,
+            offset,
+            calldata_size,
+            location,
+        ))
+        .result(0)?
+        .into();
+
+    // if offset < calldata_size => offset_ok_block
+    // else => offset_bad_block
+    ok_block.append_operation(cond_br(
+        context,
+        offset_ok,
+        &offset_ok_block,
+        &offset_bad_block,
+        &[],
+        &[],
+        location,
+    ));
+
+    /******************** offset_bad_block *******************/
+
+    // offset >= calldata_size => push 0
+    stack_push(context, &offset_bad_block, zero)?;
+    offset_bad_block.append_operation(br(&end_block, &[], location));
```

Comment:
nit. Let's leave the `cf::` prefix.
```suggestion
    offset_bad_block.append_operation(cf::br(&end_block, &[], location));
```